Task6

EvgeniyaSirenko opened the pull request:
…realisation

EvgeniyaSirenko committed:
ProductsContainer, SerializationService and SerializationServiceImpl realisation

EvgeniyaSirenko committed:
part 2 of task 6 realisation in tests and some fixes

v-zahoruiko commented on src/main/java/com/epam/preprod/sirenko/Demo.java:
```diff
@@ -1,11 +1,22 @@
 import java.io.*;
 
 public class Demo {
+	private SerializationServiceImpl serializationServiceImpl;
```
(minor) add empty line between class name and field. It can we done automatically by using crtl+alt+l bind for formatting. Rewant your other classes in this request for similar cases

EvgeniyaSirenko replied:
Kolesnikov told us not to do this because in the source code there are no empty lines between class name and field. Checked it again in ArrayList:

public class ArrayList<E> extends AbstractList<E>
        implements List<E>, RandomAccess, Cloneable, java.io.Serializable
{
    private static final long serialVersionUID = 8683452581122892189L;

Maybe it depends on project?

v-zahoruiko commented on src/main/java/com/epam/preprod/sirenko/services/SerializationService.java:
```diff
@@ -0,0 +1,7 @@
+public interface SerializationService {
+	void serializedProductsListToFile();
+	
+	void serializedProductsListOutOfFile();
```
```suggestion
	void deserializeProductsListFromFile();
```

EvgeniyaSirenko replied:
changed

v-zahoruiko commented on src/main/java/com/epam/preprod/sirenko/services/SerializationService.java:
```diff
@@ -0,0 +1,7 @@
+package com.epam.preprod.sirenko.services;
+
+public interface SerializationService {
+	void serializedProductsListToFile();
```
```suggestion
	void serializeProductsListToFile();
```

EvgeniyaSirenko replied:
changed

v-zahoruiko commented on src/main/java/com/epam/preprod/sirenko/services/impl/SerializationServiceImpl.java:
```diff
@@ -0,0 +1,48 @@
+		if (!allProducts.isEmpty()) {
+			try (
+					FileOutputStream fileOutputStream = new FileOutputStream(FILE_NAME);
+					ObjectOutputStream objectOutputStream = new ObjectOutputStream(fileOutputStream);
```
last ; can be deleted in try block

EvgeniyaSirenko replied:
done

v-zahoruiko commented on src/main/java/com/epam/preprod/sirenko/services/impl/SerializationServiceImpl.java:
```diff
@@ -0,0 +1,48 @@
+		}
+		try (
+				FileInputStream fileInputStream = new FileInputStream(FILE_NAME);
+				ObjectInputStream objectInputStream = new ObjectInputStream(fileInputStream);
```
Same

EvgeniyaSirenko replied:
done

v-zahoruiko commented on src/main/test/com/epam/preprod/sirenko/services/impl/SerializationServiceImplTest.java:
```diff
@@ -0,0 +1,78 @@
+
+import static org.junit.jupiter.api.Assertions.*;
+
+class SerializationServiceImplTest {
```
Maybe we should rename this class because in current implementation we dont test SerializationServiceImpl class here. We just serealize products by default and gzip

EvgeniyaSirenko replied:
agree, changed

EvgeniyaSirenko committed:
mistakes fixed

EvgeniyaSirenko committed:
factories classes realisation, strategy classes without realisation

EvgeniyaSirenko committed:
12.11 task6

AutoGeneratedProductImpl and CreatedByInputProductImpl realisation

v-zahoruiko commented on src/main/java/com/epam/preprod/sirenko/ClothingFactoryImpl.java:
```diff
@@ -0,0 +1,17 @@
+public class ClothingFactoryImpl implements CreateProductFactory {
+	
+	@Override
+	public Product createProduct(Strategy strategy) {
```
You shoudnt pass stratagy to factory as method param.Do it in constructor during creation

EvgeniyaSirenko replied:
done

v-zahoruiko commented on src/main/java/com/epam/preprod/sirenko/CreatedByInputProductImpl.java:
```diff
@@ -0,0 +1,126 @@
+	@Override
+	public PetGroup getPetGroup() {
+		PetGroup petGroup = null;
+		for (PetGroup element : PetGroup.values()) {
```
We can move it to separate method for better readability

EvgeniyaSirenko replied:
done

v-zahoruiko commented on src/main/java/com/epam/preprod/sirenko/CreatedByInputProductImpl.java:
```diff
@@ -0,0 +1,126 @@
+import java.io.IOException;
+import java.math.BigDecimal;
+
+public class CreatedByInputProductImpl implements Strategy {
```
Maybe we can use better naming for example ManualProductCreationStrategy

EvgeniyaSirenko replied:
done

v-zahoruiko commented on src/main/java/com/epam/preprod/sirenko/CreatedByInputProductImpl.java:
```diff
@@ -0,0 +1,126 @@
+		return price;
+	}
+	
+	private boolean checkInputIsCorrectPetGroup(String inputString) {
```
Can be moved to some class for validation as well as other check methods

EvgeniyaSirenko replied:
done

v-zahoruiko commented on src/main/java/com/epam/preprod/sirenko/Strategy.java:
```diff
@@ -0,0 +1,17 @@
+	Season getSeason();
+	Size getSize();
+	String getStringName();
+	String getStringBrandName();
```
Do we need 2 separate methods for string fields?

EvgeniyaSirenko replied:
couldn't find the way to do it with one method: in ManualProductCreationStrategy if I'll have one method I can't put two different requests like "enter name/brandName" in it

v-zahoruiko commented on src/main/java/com/epam/preprod/sirenko/command/impl/CreateNewProductCommand.java:
```diff
@@ -0,0 +1,55 @@
+			}
+			PrintToConsole.printString(CHOOSE_STRATEGY);
+			String inputStrategy = ConsoleReader.readFromConsole();
+			if (!ValidatorOfConsoleInput.checkInputStringIsNumberOneOrZero(inputStrategy)) {
```
Select the option for product creation strategy once when starting the application. You should no do it during first product creation. Also we should not change product creation strategy, but currently we can

EvgeniyaSirenko replied:
changed

v-zahoruiko commented on src/main/test/com/epam/preprod/sirenko/services/impl/SerializeProductsByDefaultAndGZipTest.java:
```diff
@@ -0,0 +1,79 @@
+	@Test
+	void shouldSaveProductsListToThreadCurrentTimes() {
+		int currentTimes = 3;
+		try (
```
We can move it to separate method for better readability

EvgeniyaSirenko replied:
done

v-zahoruiko commented on src/main/test/com/epam/preprod/sirenko/services/impl/SerializeProductsByDefaultAndGZipTest.java:
```diff
@@ -0,0 +1,79 @@
+	@Test
+	void shouldSaveProductsListAsGZip() {
+		int currentTimes = 3;
+		try (
```
Same

EvgeniyaSirenko replied:
done

EvgeniyaSirenko committed:
17.11 task6

OnProgramStartSerializeProductsListCommand realisation with serialisation part from Demo,
CommandContainer changed to send FactoryContainer to it from Demo,
Strategy field added to all Factories,
FactoryContainer field added to CreateNewProductCommand,
choose of strategy moved to Demo,
ManualProductCreationStrategy realisation of two empty methods and moved util methods to Util,
moved Strategy from ProductService

v-zahoruiko commented on src/main/java/com/epam/preprod/sirenko/Demo.java:
```diff
@@ -1,25 +1,36 @@
+			PrintToConsole.printString(INCORRECT_INPUT_ONE_OR_ZERO);
+			inputStrategy = ConsoleReader.readFromConsole();
+		}
+		Strategy strategy;
```
Can be moved to private method for better readability

EvgeniyaSirenko replied:
done

v-zahoruiko commented on src/main/java/com/epam/preprod/sirenko/command/impl/CreateNewProductCommand.java:
```diff
@@ -34,19 +30,8 @@ public void execute() {
-			}
-			productService.createNewProduct(strategy, inputCategory);
+			CreateProductFactory createProductFactory = factoryContainer.getFactory(inputCategory);
+			productService.createNewProduct(createProductFactory);
```
As we have already discussed, you should pass factory to service during service creation, but not in method for product entity creation.You can pass in in constructor or even pass factoryContainer to it instead of  just factory and get appropriate factory in service from it.So we dont need factoryContainer in this command

EvgeniyaSirenko replied:
I moved it to service but I still need factoryContainer here to print categories

v-zahoruiko commented on src/main/java/com/epam/preprod/sirenko/containers/FactoryContainer.java:
```diff
@@ -1,22 +1,28 @@
 
 public class FactoryContainer {
 	private Map<String, CreateProductFactory> factories = new HashMap<>();
+	private Strategy strategy;
```
We dont need strategy as a field here.You can pass it do directly in factories in init method

EvgeniyaSirenko replied:
I get strategy from user in Demo and need to pass strategy to FactoryContainer like this: 
FactoryContainer factoryContainer = new FactoryContainer(strategy);
Couldn't find other way to pass strategy other than in field

EvgeniyaSirenko committed:
19.11 task6

OnProgramStartSerializeProductsListCommand renamed,
separate method in Demo for readability

EvgeniyaSirenko committed:
23.11 task 6
- FactoryContainer and CreateProductFactory moved to service
- DryFoodFactoryImplTest, FoodFactoryImplTest, ClothingFactoryImplTest, AutoGeneratedProductCreationStrategyTest and ManualProductCreationStrategyTest added
- private method added to Demo for better readability
- CreateProductFactory moved from CreateNewProductCommand to service

v-zahoruiko commented on src/main/java/com/epam/preprod/sirenko/strategy/impl/AutoGeneratedProductCreationStrategy.java:
```diff
@@ -45,11 +45,11 @@ public String getStringBrandName() {
 	@Override
 	public int getInt() {
-		return RANDOM.nextInt(1000);
+		return random.nextInt(1000);
```
1000 can be moved to constants

EvgeniyaSirenko replied:
I'm using random with other nextInt(), should I make all of them constants? And it was constant but you said to get rid of static - should I make it constant again?

EvgeniyaSirenko replied:
done

v-zahoruiko commented on src/main/java/com/epam/preprod/sirenko/strategy/impl/AutoGeneratedProductCreationStrategy.java:
```diff
@@ -45,11 +45,11 @@ public String getStringBrandName() {
 	@Override
 	public BigDecimal getBigDecimal() {
-		return new BigDecimal(BigInteger.valueOf(RANDOM.nextInt(100001)), 2);
+		return new BigDecimal(BigInteger.valueOf(random.nextInt(100001)), 2);
```
Same

EvgeniyaSirenko replied:
done

v-zahoruiko commented on src/main/java/com/epam/preprod/sirenko/Demo.java:
```diff
@@ -33,6 +32,13 @@ private static void chooseStrategy() throws IOException {
+		return inputStrategy;
+	}
+	
 	public static void main(String[] args) throws IOException {
```
Move public methods before private methods

EvgeniyaSirenko replied:
done

v-zahoruiko commented on src/main/java/com/epam/preprod/sirenko/Demo.java:
```diff
@@ -33,6 +32,13 @@ private static void chooseStrategy() throws IOException {
+	}
+	
 	public static void main(String[] args) throws IOException {
 		chooseStrategy();
```
We are not only choosing strategy here. We also init command container etc. Maybe we should remove creation of FactoryContainer and CommandContainer from it, rename to getStrategy and make method return strategy instead of void

EvgeniyaSirenko replied:
Thats a good thing to remember, done

v-zahoruiko commented on src/main/test/com/epam/preprod/sirenko/strategy/impl/AutoGeneratedProductCreationStrategyTest.java:
```diff
@@ -0,0 +1,72 @@
+import static org.mockito.Mockito.when;
+
+class AutoGeneratedProductCreationStrategyTest {
+	private AutoGeneratedProductCreationStrategy autoGeneratedProduct;
```
Why your strategy is named like it is product?

EvgeniyaSirenko replied:
But it was your suggestion to name the other strategy: "maybe we can use better naming for example ManualProductCreationStrategy", so I named this one alike

v-zahoruiko replied:
Im talking about field name,not class name

EvgeniyaSirenko replied:
changed

v-zahoruiko commented on src/main/test/com/epam/preprod/sirenko/strategy/impl/AutoGeneratedProductCreationStrategyTest.java:
```diff
@@ -0,0 +1,72 @@
+	
+	@BeforeEach
+	public void SetData() {
+		autoGeneratedProduct = mock(AutoGeneratedProductCreationStrategy.class);
```
Its wrong approach to test your class. You should not mock test object. You should mock external components of your class (like some other class that is a field in your test object)

v-zahoruiko commented on src/main/test/com/epam/preprod/sirenko/strategy/impl/ManualProductCreationStrategyTest.java:
```diff
@@ -0,0 +1,74 @@
+	
+	@BeforeEach
+	public void SetData() {
+		manualProductCreation = mock(ManualProductCreationStrategy.class);
```
Same here. Also we can use better naming

EvgeniyaSirenko committed:
25.11 task 7
- chooseStrategy changed to getStrategy, other functionality moved to main method;
- constants added to AutoGeneratedProductCreationStrategy;

EvgeniyaSirenko committed:
pom file

EvgeniyaSirenko committed:
26.11 task 6
- AutoGeneratedProductCreationStrategyTest fixed: no strategy mock
- random field in AutoGeneratedProductCreationStrategy changed to not final for mock

v-zahoruiko commented on src/main/test/com/epam/preprod/sirenko/strategy/impl/AutoGeneratedProductCreationStrategyTest.java:
```diff
@@ -3,75 +3,98 @@
+	private final char[] alphabet = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ".toCharArray();
+	
+	@InjectMocks
+	AutoGeneratedProductCreationStrategy autoGeneratedProductCreationStrategy;
```
Can be private

EvgeniyaSirenko replied:
changed

v-zahoruiko commented on src/main/test/com/epam/preprod/sirenko/strategy/impl/AutoGeneratedProductCreationStrategyTest.java:
```diff
@@ -3,75 +3,98 @@
+	AutoGeneratedProductCreationStrategy autoGeneratedProductCreationStrategy;
+	
+	@Mock
+	Random random;
```
Same

EvgeniyaSirenko replied:
changed

v-zahoruiko commented on src/main/test/com/epam/preprod/sirenko/strategy/impl/AutoGeneratedProductCreationStrategyTest.java:
```diff
@@ -3,75 +3,98 @@
-	void getStringBrandName() {
-	//	when(autoGeneratedProductCreation.getStringBrandName()).thenReturn("Brand");
+	void getStringBrandNameShouldReturnString() {
+		when(random.nextInt(anyInt())).thenReturn(1).thenReturn(2).thenReturn(3).thenReturn(4).thenReturn(5)
+				.thenReturn(6).thenReturn(7).thenReturn(8).thenReturn(9).thenReturn(10);
+		StringBuilder stringBuilder = new StringBuilder();
+		for (int i = 1; i < 11; i++) {
+			char letter = alphabet[i];
+			stringBuilder.append(letter);
+		}
```
Can be moved to private methods for better readability.Also it will allow us to get rid of code duplication

EvgeniyaSirenko replied:
done

EvgeniyaSirenko committed:
30.11 task 6
 - access modifiers changed to private in AutoGeneratedProductCreationStrategyTest;
 - created private methods in AutoGeneratedProductCreationStrategyTest for better readability;
 - ConsoleReader method made non static for mocking, changed all classes that used it static;
 - ManualProductCreationStrategyTest mock logic fixed, methods names changed;
 - some methods in ManualProductCreationStrategy changed to protected for mocking;

EvgeniyaSirenko committed:
1.12 task 6
 - redundant empty lines deleted in classes;
 - getStringNameShouldReturnString method in ManualProductCreationStrategyTest fixed - string changed to pass validation;

v-zahoruiko commented on src/main/test/com/epam/preprod/sirenko/strategy/impl/ManualProductCreationStrategyTest.java:
```diff
@@ -0,0 +1,97 @@
+	
+	@BeforeEach
+	public void SetData() {
+		outMock = mock(PrintStream.class);
```
Please choose single style for mocks. Use or mock() method or annotations for consistency

EvgeniyaSirenko replied:
done

v-zahoruiko commented on src/main/test/com/epam/preprod/sirenko/strategy/impl/ManualProductCreationStrategyTest.java:
```diff
@@ -0,0 +1,97 @@
+	private ConsoleReader consoleReader;
+	
+	@BeforeEach
+	public void SetData() {
```
```suggestion
	public void setup() {
```

EvgeniyaSirenko replied:
changed

v-zahoruiko commented on src/main/test/com/epam/preprod/sirenko/strategy/impl/ManualProductCreationStrategyTest.java:
```diff
@@ -0,0 +1,97 @@
+	@Test
+	void getPetGroupShouldReturnPetGroup() throws IOException {
+		doNothing().when(manualProductCreationStrategy).printPetGroupElements();
+		doNothing().when(outMock).println();
```
doNothing().when(outMock).println(); is duplicated in each scenario

EvgeniyaSirenko replied:
made private method

EvgeniyaSirenko committed:
2.12 task 6

changes in ManualProductCreationStrategyTest:

- created a private method mockPrintln for better readability;
- changed SetData tp setup;
- changed @Mock to mock for all classes;
